In [1]:
import numpy as np
import pandas as pd
import time
from workalendar.asia import SouthKorea
#import MySQLdb
import sqlalchemy
from datetime import datetime, timedelta, date
import pymysql as mdb
mdb.install_as_MySQLdb()
import sqlalchemy
from sqlalchemy import create_engine
from sqlalchemy.sql import text

#### NOTE: you might need to install this library:
#### pip install workalendar

# Extract 

###  Extract CSV files

In [2]:
# Convert a csv file to DataFrame (pandas object)
file_name = 'patient.csv'
patient_df = pd.read_csv(file_name)

file_name = 'region.csv'
region_df = pd.read_csv(file_name)

file_name = 'route.csv'
route_df = pd.read_csv(file_name)

file_name = 'time.csv'
time_df = pd.read_csv(file_name)


## Patient Extract&Transform

In [3]:
patient_df.head()

patient_id  global_num     sex  birth_year  age country province  \
0  1000000001         2.0    male      1964.0  50s   Korea    Seoul   
1  1000000002         5.0    male      1987.0  30s   Korea    Seoul   
2  1000000003         6.0    male      1964.0  50s   Korea    Seoul   
3  1000000004         7.0    male      1991.0  20s   Korea    Seoul   
4  1000000005         9.0  female      1992.0  20s   Korea    Seoul   

          city disease        infection_case  infection_order   infected_by  \
0   Gangseo-gu     NaN       overseas inflow              1.0           NaN   
1  Jungnang-gu     NaN       overseas inflow              1.0           NaN   
2    Jongno-gu     NaN  contact with patient              2.0  2.002000e+09   
3      Mapo-gu     NaN       overseas inflow              1.0           NaN   
4  Seongbuk-gu     NaN  contact with patient              2.0  1.000000e+09   

   contact_number symptom_onset_date confirmed_date released_date  \
0            75.0         2020-01-22     2020-01-23    2020-02-05   
1            31.0                NaN     2020-01-30    2020-03-02   
2            17.0                NaN     2020-01-30    2020-02-19   
3             9.0         2020-01-26     2020-01-30    2020-02-15   
4             2.0                NaN     2020-01-31    2020-02-24   

  deceased_date     state  
0           NaN  released  
1           NaN  released  
2           NaN  released  
3           NaN  released  
4           NaN  released

In [4]:
#Table shape
dim_patient = patient_df.shape
print("we have {} rows and {} columns in patient Data ".format(dim_patient[0],dim_patient[1]))

we have 2243 rows and 18 columns in patient Data 


In [5]:
# Retrieving fields
fields_patient = patient_df.columns
print('fields_patient: {}'.format(fields_patient))

fields_patient: Index(['patient_id', 'global_num', 'sex', 'birth_year', 'age', 'country',
       'province', 'city', 'disease', 'infection_case', 'infection_order',
       'infected_by', 'contact_number', 'symptom_onset_date', 'confirmed_date',
       'released_date', 'deceased_date', 'state'],
      dtype='object')


In [6]:
# Slicing the relevant columns
patient_df_2 = patient_df.copy()
patient_df_2 = patient_df_2.loc[:,['patient_id','birth_year', 'province','infected_by','confirmed_date','released_date','deceased_date','state']]


In [7]:
# Learning types
patient_df_2.dtypes

patient_id          int64
birth_year        float64
province           object
infected_by       float64
confirmed_date     object
released_date      object
deceased_date      object
state              object
dtype: object

In [8]:
# Changing data types
patient_df_2['confirmed_date'] = pd.to_datetime(patient_df_2['confirmed_date'])
patient_df_2['released_date'] = pd.to_datetime(patient_df_2['released_date'])
patient_df_2['deceased_date'] = pd.to_datetime(patient_df_2['deceased_date'])

patient_df_2.dtypes


patient_id                 int64
birth_year               float64
province                  object
infected_by              float64
confirmed_date    datetime64[ns]
released_date     datetime64[ns]
deceased_date     datetime64[ns]
state                     object
dtype: object

In [9]:
# Changing province and state to lower case 
patient_df_2['province'] = patient_df_2['province'].str.lower()
patient_df_2['state'] = patient_df_2['state'].str.lower()

patient_df_2.head()

patient_id  birth_year province   infected_by confirmed_date released_date  \
0  1000000001      1964.0    seoul           NaN     2020-01-23    2020-02-05   
1  1000000002      1987.0    seoul           NaN     2020-01-30    2020-03-02   
2  1000000003      1964.0    seoul  2.002000e+09     2020-01-30    2020-02-19   
3  1000000004      1991.0    seoul           NaN     2020-01-30    2020-02-15   
4  1000000005      1992.0    seoul  1.000000e+09     2020-01-31    2020-02-24   

  deceased_date     state  
0           NaT  released  
1           NaT  released  
2           NaT  released  
3           NaT  released  
4           NaT  released

In [10]:
# Check for null values
patient_df_2.isnull().sum()

patient_id           0
birth_year         454
province             0
infected_by       1768
confirmed_date     145
released_date     2010
deceased_date     2213
state               96
dtype: int64

In [11]:
# calculating age for patient

# filling the null birth-year objects with median
year_median = patient_df_2['birth_year'].median()
patient_df_2['birth_year'].fillna(year_median, inplace = True)

# calc the current year
current_year = pd.datetime.now().year

# calc the age for all patients, creating a new column for age
patient_df_2['age'] = current_year-patient_df_2['birth_year']

patient_df_2['age'].head()

C:\Users\Paz\anaconda3\lib\site-packages\ipykernel_launcher.py:8: FutureWarning: The pandas.datetime class is deprecated and will be removed from pandas in a future version. Import from datetime module instead.
  


0    56.0
1    33.0
2    56.0
3    29.0
4    28.0
Name: age, dtype: float64

In [12]:
# Filling null for confirmed_date
patient_df_2['confirmed_date'].fillna(method = "pad", inplace = True) 

In [13]:
# Adding mean relese date where relese date null and status = 'released'

df = patient_df_2[(patient_df_2['released_date'].notna()) & (patient_df_2['state'] == 'released')]
deasease_duration = (df['released_date'] - df['confirmed_date']).mean().days
patient_df_2.loc[(patient_df_2['state'] == 'released')& (patient_df_2['released_date'].isnull()), 'released_date'] = patient_df_2['confirmed_date'] + pd.DateOffset(days = int(deasease_duration)) 


In [14]:
# Verfying isolated status count
patient_df_2[patient_df_2['state'] == 'isolated'].count()

patient_id        1801
birth_year        1801
province          1801
infected_by        317
confirmed_date    1801
released_date        3
deceased_date        0
state             1801
age               1801
dtype: int64

###### we can see there is 3 people with status "isolated" but with "released date"

In [15]:
# Fixing the problem above by changing the state from isolated --> released

patient_df_2.loc[(patient_df_2['state'] =='isolated') & (patient_df_2['released_date'].notna()), 'state'] = 'released'
print ('{} patients with status isolated but with released date'.format(patient_df_2[patient_df_2['state']=='isolated'].count().released_date))


0 patients with status isolated but with released date


In [16]:
# Cheaking if the number of released date is the same as "released" status

print("The total released status: {}".format(str(patient_df_2[patient_df_2['state']=='released'].count().state)))
print("The total released dates: {}".format(str(patient_df_2['released_date'].count())))

The total released status: 317
The total released dates: 317


In [17]:
# Cheaking if there are null states that has releaesed or deceased dates in order to fill it if nesscery
no_status_released = patient_df_2[(patient_df_2['state'].isnull()) & (patient_df_2['released_date'].notna())].state.count()
no_status_dead = patient_df_2[(patient_df_2['state'].isnull()) & (patient_df_2['deceased_date'].notna())].state.count()

print("Number of patient with no state and have released date: "+ str(no_status_released))
print("Number of patient with no state and have deceased_date: "+ str(no_status_dead))


Number of patient with no state and have released date: 0
Number of patient with no state and have deceased_date: 0


In [18]:
# Because we don't have any dates ("relsead_date/deceased_date") for missing state values, we will assume they are isolated and fill the state null with isolated value
patient_df_2.loc[(patient_df_2['state'].isnull()) , 'state'] = 'isolated'


In [19]:
# Cheacking if there is a problem with the num of null values in realesead date
no_release_date = patient_df_2.isnull().sum().released_date
isoleted = patient_df_2[patient_df_2['state'] == 'isolated'].count().patient_id
deceased = patient_df_2[patient_df_2['state'] == 'deceased'].count().patient_id

print ('The number of patients with no release date: {}, The number of isoleted patient is: {},The number of deceased patients is: {}'.format((no_release_date),(isoleted),(deceased)))


The number of patients with no release date: 1926, The number of isoleted patient is: 1894,The number of deceased patients is: 32


###### 1926 null for released date is fine, because part of the patients are still isolated and part of them have died.

In [20]:
# Checking if all realese date is after confirmed date
patient_df_2[patient_df_2['released_date'] < patient_df_2['confirmed_date']].count()

patient_id        0
birth_year        0
province          0
infected_by       0
confirmed_date    0
released_date     0
deceased_date     0
state             0
age               0
dtype: int64

In [21]:
# Checking if null values left to handle
patient_df_2.isnull().sum()

patient_id           0
birth_year           0
province             0
infected_by       1768
confirmed_date       0
released_date     1926
deceased_date     2213
state                0
age                  0
dtype: int64

##### Yes there are but it's fine cause not everyone recoverd, alive or infected someone.
##### Let's check the infected_by column and fill the null values:

In [22]:
# Fill the null values with 0
patient_df_2.loc[(patient_df_2['infected_by'].isnull()), 'infected_by'] = 0

In [23]:
# looking for outliers, using IQR 
Q1 = patient_df_2['age'].quantile(0.25)
Q3 = patient_df_2['age'].quantile(0.75)
IQR = Q3 - Q1
print(IQR)

23.0


In [24]:
wsub = patient_df_2['age']
X = (wsub < (Q1 - 2 * IQR)) | (wsub > (Q3 + 2* IQR))

patient_df_2[X]

patient_id  birth_year          province  infected_by confirmed_date  \
1602  6001000503      1916.0  gyeongsangbuk-do          0.0     2020-03-10   

     released_date deceased_date     state    age  
1602           NaT           NaT  isolated  104.0

In [25]:
# def for changing type to int
def replace_int(x):
    return int(x)

In [26]:
# Change to type -> int
patient_df_2["infected_by"] = patient_df_2["infected_by"].apply(replace_int)
patient_df_2['age'] = patient_df_2['age'].apply(replace_int)


##### we assume that 104 is not an outlier,  so we won't remove it

In [27]:
# Func that return 1 if the patient is relsead otherwise, 0.
def is_released_to_boolean(date):
    if date == 'nan':
        return 0
    else:
        return 1

In [28]:
# Add a new column to patient_df, for patient is released it will get 1, otherwise 0. 
patient_df_2['state'] = patient_df_2['released_date'].apply(is_released_to_boolean)
patient_df_2.rename(columns={'state': 'is_released'}, inplace=True)

patient_df_2.head()

patient_id  birth_year province  infected_by confirmed_date released_date  \
0  1000000001      1964.0    seoul            0     2020-01-23    2020-02-05   
1  1000000002      1987.0    seoul            0     2020-01-30    2020-03-02   
2  1000000003      1964.0    seoul   2002000001     2020-01-30    2020-02-19   
3  1000000004      1991.0    seoul            0     2020-01-30    2020-02-15   
4  1000000005      1992.0    seoul   1000000002     2020-01-31    2020-02-24   

  deceased_date  is_released  age  
0           NaT            1   56  
1           NaT            1   33  
2           NaT            1   56  
3           NaT            1   29  
4           NaT            1   28

#### Part of the deceased_date column is with null values because, not all patients has deceased. We didn't dill with it because we won't  use it  in any feuture calculation.

## Dim Patient

In [29]:
# Creating DIM PATIENT 

# Slicing the relevant columns
dim_patient = patient_df_2.loc[:,['patient_id','age','province']]

# Dropping duplicates if there are any
dim_patient.drop_duplicates()

# Creating Auto-increment column for dim patient
dim_patient['patient_key'] = np.arange(dim_patient.shape[0])+1

# Arrange the columns in the right order
dim_patient = dim_patient[['patient_key',"patient_id", "age",'province']]

print ("The types of dim patient are:")
print (dim_patient.dtypes)

shape_dim_patient = dim_patient.shape
print(" \nWe have {} rows and {} columns in dim patient \n".format(shape_dim_patient[0],shape_dim_patient[1]))

print("The range of ages is: ({}, {}) \n".format ((dim_patient.age.agg(["min","max"]).min()),(dim_patient.age.agg(["min","max"]).max())))

print(dim_patient.head())

The types of dim patient are:
patient_key     int32
patient_id      int64
age             int64
province       object
dtype: object
 
We have 2243 rows and 4 columns in dim patient 

The range of ages is: (0, 104) 

   patient_key  patient_id  age province
0            1  1000000001   56    seoul
1            2  1000000002   33    seoul
2            3  1000000003   56    seoul
3            4  1000000004   29    seoul
4            5  1000000005   28    seoul


## Route EXtract&Transform

In [30]:
route_df_2 = route_df.copy()
route_df_2.head()

patient_id  global_num        date     province      city   latitude  \
0  1400000001           1  2020-01-19      Incheon   Jung-gu  37.460459   
1  1400000001           1  2020-01-20      Incheon    Seo-gu  37.478832   
2  1000000001           2  2020-01-22  Gyeonggi-do  Gimpo-si  37.562143   
3  1000000001           2  2020-01-23        Seoul   Jung-gu  37.567454   
4  2000000001           3  2020-01-20      Incheon   Jung-gu  37.460459   

    longitude  
0  126.440680  
1  126.668558  
2  126.801884  
3  127.005627  
4  126.440680

In [31]:
#Learning types
route_df_2.dtypes

patient_id      int64
global_num      int64
date           object
province       object
city           object
latitude      float64
longitude     float64
dtype: object

In [32]:
# Check for null values
route_df_2.isnull().sum()

patient_id    0
global_num    0
date          0
province      0
city          0
latitude      0
longitude     0
dtype: int64

In [33]:
# Table shape
shape_route_df_2 = route_df_2.shape
print("We have {} rows and {} columns in route_df_2".format(shape_route_df_2[0],shape_route_df_2[1]))

We have 175 rows and 7 columns in route_df_2


In [34]:
# Retrieving fields
fields_route = route_df_2.columns
print('fields_route: {}'.format(fields_route))

fields_route: Index(['patient_id', 'global_num', 'date', 'province', 'city', 'latitude',
       'longitude'],
      dtype='object')


In [35]:
# Slicing the relevant columns
route_df_2 = route_df_2.loc[:,['patient_id','province','date']]

In [36]:
# Change the type from object to date in route df 'date' coulmn
route_df_2['date'] = pd.to_datetime(route_df_2['date'])

# Changing province to lower case 
route_df_2['province'] = route_df_2['province'].str.lower()

route_df_2.dtypes

patient_id             int64
province              object
date          datetime64[ns]
dtype: object

##### Preparing the data for the aggregative columns in fact table

In [37]:
# Calculating the places patient visited "number_of_days" before "confirmed_date", the day of isolation
def calc_days(x,number_of_days):
    return route_df_2[(route_df_2['patient_id'] == x['patient_id']) & route_df_2.date.between((x['confirmed_date'] - pd.DateOffset(days = int(number_of_days))), (x['confirmed_date']-pd.DateOffset(days = int(1))))].shape[0]


In [38]:
# Applying the calculations for 1,2,3 days before isolation
patient_df_2['visited_1'] = patient_df_2.apply(lambda x: calc_days(x,1), axis=1)
patient_df_2['visited_2'] = patient_df_2.apply(lambda x: calc_days(x,2), axis=1)
patient_df_2['visited_3'] = patient_df_2.apply(lambda x: calc_days(x,3), axis=1)

patient_df_2.head()

patient_id  birth_year province  infected_by confirmed_date released_date  \
0  1000000001      1964.0    seoul            0     2020-01-23    2020-02-05   
1  1000000002      1987.0    seoul            0     2020-01-30    2020-03-02   
2  1000000003      1964.0    seoul   2002000001     2020-01-30    2020-02-19   
3  1000000004      1991.0    seoul            0     2020-01-30    2020-02-15   
4  1000000005      1992.0    seoul   1000000002     2020-01-31    2020-02-24   

  deceased_date  is_released  age  visited_1  visited_2  visited_3  
0           NaT            1   56          1          1          1  
1           NaT            1   33          1          5          5  
2           NaT            1   56          0          0          0  
3           NaT            1   29          0          0          0  
4           NaT            1   28          0          0          0

## Region Extract&Transform


In [39]:
region_df_2 = region_df.copy()
region_df_2.head()

code province         city   latitude   longitude  \
0  10000    Seoul        Seoul  37.566953  126.977977   
1  10010    Seoul   Gangnam-gu  37.518421  127.047222   
2  10020    Seoul  Gangdong-gu  37.530492  127.123837   
3  10030    Seoul   Gangbuk-gu  37.639938  127.025508   
4  10040    Seoul   Gangseo-gu  37.551166  126.849506   

   elementary_school_count  kindergarten_count  university_count  \
0                      607                 830                48   
1                       33                  38                 0   
2                       27                  32                 0   
3                       14                  21                 0   
4                       36                  56                 1   

   academy_ratio  elderly_population_ratio  elderly_alone_ratio  \
0           1.44                     15.38                  5.8   
1           4.18                     13.17                  4.3   
2           1.54                     14.55                  5.4   
3           0.67                     19.49                  8.5   
4           1.17                     14.39                  5.7   

   nursing_home_count  
0               22739  
1                3088  
2                1023  
3                 628  
4                1080

In [40]:
#Learning types
region_df_2.dtypes

code                          int64
province                     object
city                         object
latitude                    float64
longitude                   float64
elementary_school_count       int64
kindergarten_count            int64
university_count              int64
academy_ratio               float64
elderly_population_ratio    float64
elderly_alone_ratio         float64
nursing_home_count            int64
dtype: object

In [41]:
# Table shape
shape_region_df_2 = region_df_2.shape
print("We have {} rows and {} columns in region_df_2".format(shape_region_df_2[0],shape_region_df_2[1]))


We have 244 rows and 12 columns in region_df_2


In [42]:
# Check for null values
region_df_2.isnull().sum()

code                        0
province                    0
city                        0
latitude                    0
longitude                   0
elementary_school_count     0
kindergarten_count          0
university_count            0
academy_ratio               0
elderly_population_ratio    0
elderly_alone_ratio         0
nursing_home_count          0
dtype: int64

In [43]:
# Retrieving fields
fields_region = region_df_2.columns
print('fields_region: {}'.format(fields_region))

fields_region: Index(['code', 'province', 'city', 'latitude', 'longitude',
       'elementary_school_count', 'kindergarten_count', 'university_count',
       'academy_ratio', 'elderly_population_ratio', 'elderly_alone_ratio',
       'nursing_home_count'],
      dtype='object')


In [44]:
# Check the number of provinces we have
print('The number of provinces in region table is : {}'.format(region_df_2['province'].nunique()))
# Checking that the number of provinces in patient table isn't higher from the one in region table.
print('The number of provinces in patient table is: {}'.format(patient_df_2['province'].nunique()))

The number of provinces in region table is : 18
The number of provinces in patient table is: 17


In [45]:
# Slicing to use only the necessary coulmns & remove duplicate rows
region_df_2 = region_df_2.loc[:,['province']].drop_duplicates()

# Changing province to lower case 
region_df_2['province'] = region_df_2['province'].str.lower()

region_df_2.head()

province
0     seoul
26    busan
43    daegu
52  gwangju
58  incheon

#### Preparing the data for the aggregative columns in dim_place

In [46]:
# Calc the top 3 provinces with the highest number of infected patients
top_3_province = patient_df_2.groupby('province').size().sort_values(ascending=False).head(3)
print (top_3_province)

to_list_top_3_province = top_3_province.reset_index()['province'].tolist()

# Define the provinces with the highest number of infected patients to boolean func
def top_3_to_boolean(x):
    if x in to_list_top_3_province:
        return 1
    else:
        return 0

province
gyeongsangbuk-do    1054
gyeonggi-do          300
seoul                299
dtype: int64


In [47]:
# Adding a new coulmn: Top 3 - the top 3 provinces with the highest number of infected patients
region_df_2['top_3'] = region_df_2['province'].apply(top_3_to_boolean)
region_df_2.head()

province  top_3
0     seoul      1
26    busan      0
43    daegu      0
52  gwangju      0
58  incheon      0

In [48]:
# Calc the total infected patients for each province and join it to region df
Total_patients_by_province = patient_df_2.groupby('province').size().reset_index()

# Rename the column
Total_patients_by_province.rename( columns={0:'total_patients'}, inplace=True )

# join the two df 
region_df_2 = pd.merge(region_df_2, Total_patients_by_province, how ='left', validate = 'one_to_one')

print(region_df_2.head())

  province  top_3  total_patients
0    seoul      1           299.0
1    busan      0           101.0
2    daegu      0            63.0
3  gwangju      0            19.0
4  incheon      0            39.0


In [49]:
# Adding a new coulmn - Auto increment: Place_id
region_df_2['place_id'] = np.arange(region_df_2.shape[0]) + 1

# Fill the null values 
region_df_2.fillna(0, inplace = True)

# using func replace_int to change to type int
region_df_2["total_patients"] = region_df_2["total_patients"].apply(replace_int)

# arranging the columns on the right order
region_df_2 = region_df_2[['place_id','province','total_patients','top_3']]
region_df_2.head()

place_id province  total_patients  top_3
0         1    seoul             299      1
1         2    busan             101      0
2         3    daegu              63      0
3         4  gwangju              19      0
4         5  incheon              39      0

## Dim Place

In [50]:
# Creating dim place

dim_place = region_df_2.copy()

print ("The types of dim place are:")
print (dim_place.dtypes)

shape_dim_place = dim_place.shape
print(" \nWe have {} rows and {} columns in dim place \n".format(shape_dim_place[0],shape_dim_place[1]))

print("The range of total patients is: ({}, {}) \n".format ((dim_place.total_patients.agg(["min","max"]).min()),(dim_place.total_patients.agg(["min","max"]).max())))


print(dim_place.head())

The types of dim place are:
place_id           int32
province          object
total_patients     int64
top_3              int64
dtype: object
 
We have 18 rows and 4 columns in dim place 

The range of total patients is: (0, 1054) 

   place_id province  total_patients  top_3
0         1    seoul             299      1
1         2    busan             101      0
2         3    daegu              63      0
3         4  gwangju              19      0
4         5  incheon              39      0


## Time Extract&Transform

In [51]:
time_df_2 = time_df.copy()
time_df_2.head()

date  time  test  negative  confirmed  released  deceased
0  2020-01-20    16     1         0          1         0         0
1  2020-01-21    16     1         0          1         0         0
2  2020-01-22    16     4         3          1         0         0
3  2020-01-23    16    22        21          1         0         0
4  2020-01-24    16    27        25          2         0         0

In [52]:
#Learning types
time_df_2.dtypes

date         object
time          int64
test          int64
negative      int64
confirmed     int64
released      int64
deceased      int64
dtype: object

In [53]:
# Change type object --> date
time_df_2['date'] = pd.to_datetime(time_df_2['date'])

time_df_2.dtypes

date         datetime64[ns]
time                  int64
test                  int64
negative              int64
confirmed             int64
released              int64
deceased              int64
dtype: object

In [54]:
# Table shape
shape_time_df_2 = time_df_2.shape
print("We have {} rows and {} columns in time df".format(shape_time_df_2[0],shape_time_df_2[1]))

We have 63 rows and 7 columns in time df


In [55]:
# Check for null values
time_df_2.isnull().sum()

date         0
time         0
test         0
negative     0
confirmed    0
released     0
deceased     0
dtype: int64

In [56]:
# Retrieving fields
fields_time = time_df_2.columns
print('fields_time: {}'.format(fields_time))

fields_time: Index(['date', 'time', 'test', 'negative', 'confirmed', 'released',
       'deceased'],
      dtype='object')


In [57]:
# Check the min and max dates from time table and add 14 days before the min and after the max date
# Check what is the delta between the dates (Start_date&End_date),and add all dates between those two date to 'date' column.

extreme_dates = time_df_2.date.agg(["min","max"])
Start_date = extreme_dates.min() - timedelta(days=14)
End_date = extreme_dates.max() + timedelta(days=14)
Delta = End_date - Start_date

date_list = []
for i in range(Delta.days + 1):
    day = Start_date + timedelta(days = i)
    date_list.append(day)


In [58]:
# Create dim date
dim_date = pd.DataFrame(columns=["date_key","date"])   
dim_date['date']= pd.Series(date_list)

# Add date_key auto increment
dim_date['date_key'] = np.arange(dim_date.shape[0])+1

dim_date.head()

date_key       date
0         1 2020-01-06
1         2 2020-01-07
2         3 2020-01-08
3         4 2020-01-09
4         5 2020-01-10

In [59]:
# Add new columns by using the date column

# Returning the digit of day, month, year of each date
dim_date['day'] = dim_date['date'].map(lambda x : x.day)
dim_date['month'] = dim_date['date'].map(lambda y : y.month)
dim_date['year'] = dim_date['date'].map(lambda z : z.year)

# Attaching names to days
dim_date['day_of_week'] = dim_date['date'].map(lambda m : m.day_name())


In [60]:
# Indicator of holidays, if the date is an holiday it will get 1, otherwise 0.
cal = SouthKorea()
def holiday_to_boolean (date):
    if cal.is_holiday(date):
        return 1
    else:
        return 0

In [61]:
# Appling the func above into new column
dim_date["is_holiday"] = dim_date["date"].apply(holiday_to_boolean)
dim_date.head()

date_key       date  day  month  year day_of_week  is_holiday
0         1 2020-01-06    6      1  2020      Monday           0
1         2 2020-01-07    7      1  2020     Tuesday           0
2         3 2020-01-08    8      1  2020   Wednesday           0
3         4 2020-01-09    9      1  2020    Thursday           0
4         5 2020-01-10   10      1  2020      Friday           0

In [62]:
print ("The types of dim date are:")
print (dim_date.dtypes)

shape_dim_date = dim_date.shape
print(" \nWe have {} rows and {} columns in dim date \n".format(shape_dim_date[0],shape_dim_date[1]))

print("The range of dates is: ({}, {}) \n".format ((dim_date.date.agg(["min","max"]).min()),(dim_date.date.agg(["min","max"]).max())))

print(dim_date.head())

The types of dim date are:
date_key                int32
date           datetime64[ns]
day                     int64
month                   int64
year                    int64
day_of_week            object
is_holiday              int64
dtype: object
 
We have 91 rows and 7 columns in dim date 

The range of dates is: (2020-01-06 00:00:00, 2020-04-05 00:00:00) 

   date_key       date  day  month  year day_of_week  is_holiday
0         1 2020-01-06    6      1  2020      Monday           0
1         2 2020-01-07    7      1  2020     Tuesday           0
2         3 2020-01-08    8      1  2020   Wednesday           0
3         4 2020-01-09    9      1  2020    Thursday           0
4         5 2020-01-10   10      1  2020      Friday           0


## Fact - Infection

In [63]:
# Take the relevant columns from patient df for fact df
fact_df = patient_df_2[['patient_id','is_released','province','confirmed_date','released_date','infected_by','visited_1','visited_2','visited_3']]

# Checking the shape of df
shape_fact_df = fact_df.shape
print("We have {} rows and {} columns in fact_df".format(shape_fact_df[0],shape_fact_df[1]))


We have 2243 rows and 9 columns in fact_df


In [64]:
# Merge the fact_df with dim_place in order to get the key -> place_id
fact_df = pd.merge(fact_df, dim_place, how='left',left_on = 'province', right_on = 'province', validate = 'm:1')

# Take the relevant columns
fact_df = fact_df[['patient_id','place_id','is_released','confirmed_date','released_date','visited_1','visited_2','visited_3','infected_by']]

# Checking the shape of df
shape_fact_df = fact_df.shape
print("We have {} rows and {} columns in fact_df".format(shape_fact_df[0],shape_fact_df[1]))

We have 2243 rows and 9 columns in fact_df


In [65]:
# Merge dim_date on confirmed_date in order to get the key -> date key
fact_df = pd.merge(fact_df, dim_date, how='left',left_on = 'confirmed_date', right_on = 'date', validate = 'm:1')

fact_df.rename(columns={'date_key': 'confirmed_date_key'}, inplace=True)

# Take the relevant columns
fact_df = fact_df[['patient_id','place_id','is_released','confirmed_date_key','released_date','visited_1','visited_2','visited_3','infected_by']]

# Checking the shape of df
shape_fact_df = fact_df.shape
print("We have {} rows and {} columns in fact_df".format(shape_fact_df[0],shape_fact_df[1]))


We have 2243 rows and 9 columns in fact_df


In [66]:
# Merge dim_date on released_date in order to get the key -> date key
fact_df = pd.merge(fact_df, dim_date, how='left',left_on = 'released_date', right_on = 'date', validate = 'm:1')

fact_df.rename(columns={'date_key': 'released_date_key'}, inplace=True)

# Fill the null values 
fact_df['released_date_key'].fillna(0, inplace = True)

# Using func replace_int to change to type int
fact_df["released_date_key"] = fact_df["released_date_key"].apply(replace_int)

# Take the relevant columns
fact_df = fact_df[['patient_id','place_id','is_released','confirmed_date_key','released_date_key','visited_1','visited_2','visited_3','infected_by']]

# Checking the shape of df
shape_fact_df = fact_df.shape
print("We have {} rows and {} columns in fact_df".format(shape_fact_df[0],shape_fact_df[1]))


We have 2243 rows and 9 columns in fact_df


In [67]:
# Merge dim_patient on patient_id in order to get the key -> patient key
fact_df = pd.merge(fact_df, dim_patient, how='left',left_on = 'patient_id', right_on = 'patient_id', validate = "m:1")

# Take the relevant columns
fact_df = fact_df[['patient_key','place_id','is_released','confirmed_date_key','released_date_key','visited_1','visited_2','visited_3','infected_by']]

# Checking the shape of df
shape_fact_df = fact_df.shape
print("We have {} rows and {} columns in fact_df".format(shape_fact_df[0],shape_fact_df[1]))

We have 2243 rows and 9 columns in fact_df


In [68]:
# Merge dim_patient on patient_id in order to get the key -> patient key
fact_df = pd.merge(fact_df, dim_patient, how='left',left_on = 'infected_by', right_on = 'patient_id', validate = "m:1")

fact_df.rename(columns={'patient_key_y': 'infected_by_patient_key'}, inplace=True)
fact_df.rename(columns={'patient_key_x': 'patient_key'}, inplace=True)

# Fill the null values 
fact_df['infected_by_patient_key'].fillna(0, inplace = True)

# Using func replace_int to change to type int
fact_df['infected_by_patient_key'] = fact_df['infected_by_patient_key'].apply(replace_int)

# Take the relevant columns
fact_df = fact_df[['patient_key','place_id','is_released','confirmed_date_key','released_date_key','visited_1','visited_2','visited_3','infected_by_patient_key']]

# Checking the shape of df
shape_fact_df = fact_df.shape
print("We have {} rows and {} columns in fact_df".format(shape_fact_df[0],shape_fact_df[1]))

We have 2243 rows and 9 columns in fact_df


In [69]:
# Sorting the values by confiemrd date
fact_df.sort_values(by = 'confirmed_date_key')

# Add infection_key auto increment.
fact_df['infection_key'] = np.arange(fact_df.shape[0])+1

# Arrange the columns in the right order
fact_df = fact_df[['infection_key','patient_key','place_id','is_released','confirmed_date_key','released_date_key','visited_1','visited_2','visited_3','infected_by_patient_key']]

In [70]:
# Checking types, shapes, null values
print ("The types of fact-df are:")
print (fact_df.dtypes)

print('\n{} null values in fact df'.format(fact_df.isnull().sum().sum()))

shape_fact_df = fact_df.shape
print("\nWe have {} rows and {} columns in fact df".format(shape_fact_df[0],shape_fact_df[1]))

The types of fact-df are:
infection_key              int32
patient_key                int32
place_id                   int32
is_released                int64
confirmed_date_key         int32
released_date_key          int64
visited_1                  int64
visited_2                  int64
visited_3                  int64
infected_by_patient_key    int64
dtype: object

0 null values in fact df

We have 2243 rows and 10 columns in fact df


In [71]:
fact_df.head()

infection_key  patient_key  place_id  is_released  confirmed_date_key  \
0              1            1         1            1                  18   
1              2            2         1            1                  25   
2              3            3         1            1                  25   
3              4            4         1            1                  25   
4              5            5         1            1                  26   

   released_date_key  visited_1  visited_2  visited_3  infected_by_patient_key  
0                 31          1          1          1                        0  
1                 57          1          5          5                        0  
2                 45          0          0          0                        0  
3                 41          0          0          0                        0  
4                 50          0          0          0                        2

## Load to Mysql

In [72]:
engine = create_engine("mysql://root:root@localhost")
con = engine.connect()

C:\Users\Paz\anaconda3\lib\site-packages\pymysql\cursors.py:170: Warning: (1366, "Incorrect string value: '\\xF9\\xF2\\xE5\\xEF \\xF7...' for column 'VARIABLE_VALUE' at row 519")
  result = self._query(query)


In [73]:
#create database covid_19
con.execute('DROP SCHEMA IF EXISTS covid_19')
con.execute('CREATE SCHEMA IF NOT EXISTS covid_19')

In [74]:
#use database covid_19
con.execute('USE covid_19')

#create table - dim patient
con.execute("""CREATE TABLE covid_19.dim_patient 
           (patient_key INT AUTO_INCREMENT PRIMARY KEY, 
            patient_id BIGINT,
            age INT,
            province VARCHAR(45) )""")

In [75]:
#create the table - dim place
con.execute("""CREATE TABLE covid_19.dim_place 
           (place_id INT AUTO_INCREMENT PRIMARY KEY, 
            province VARCHAR(45), 
            total_patients INT, 
            top_3 TINYINT)""")

In [76]:
#create the table - dim date
con.execute("""CREATE TABLE covid_19.dim_date
           (date_key INT AUTO_INCREMENT PRIMARY KEY, 
            date DATE, 
            day INT,
            month INT,
            year INT,
            day_of_week VARCHAR(11),
            is_holiday TINYINT)""")

In [77]:
#create the table - fact infection
con.execute("""CREATE TABLE covid_19.fact_df 
           (infection_key INT AUTO_INCREMENT PRIMARY KEY, 
            patient_key INT, 
            place_id INT,
            is_released TINYINT,
            confirmed_date_key INT,
            released_date_key INT,
            visited_1 INT,
            visited_2 INT,
            visited_3 INT,
            infected_by_patient_key INT)""")

#con.execute('SET SQL_MODE=ANSI_QUOTES') 

In [78]:
# Create and use patient dimaention in MySql
dim_patient.to_sql(con=con, schema='covid_19', name='dim_patient', if_exists='append',index=False)

In [79]:
# Create and use place dimaention in MySql
dim_place.to_sql(con=con, schema='covid_19', name='dim_place', if_exists='append',index=False)

In [80]:
# Create and use date dimaention in MySql
dim_date.to_sql(con=con, schema='covid_19', name='dim_date', if_exists='append',index=False)

In [81]:
# Create and use infection fact table in MySql
fact_df.to_sql(con=con, schema='covid_19', name='fact_df', if_exists='append',index=False)

### Checking if we load the data currectly

In [82]:
dim_patient_table = pd.read_sql("SELECT * from covid_19.dim_patient", con)
dim_patient_table.head()

patient_key  patient_id  age province
0            1  1000000001   56    seoul
1            2  1000000002   33    seoul
2            3  1000000003   56    seoul
3            4  1000000004   29    seoul
4            5  1000000005   28    seoul

In [83]:
# Checking the shape of table
shape_dim_patient_table= dim_patient_table.shape
print("We have {} rows and {} columns in dim_patient_table".format(shape_dim_patient_table[0],shape_dim_patient_table[1]))

We have 2243 rows and 4 columns in dim_patient_table


In [84]:
dim_place_table = pd.read_sql("SELECT * from covid_19.dim_place", con)
dim_place_table.head()

place_id province  total_patients  top_3
0         1    seoul             299      1
1         2    busan             101      0
2         3    daegu              63      0
3         4  gwangju              19      0
4         5  incheon              39      0

In [85]:
# Checking the shape of table
shape_dim_place_table = dim_place_table.shape
print("We have {} rows and {} columns in dim_place_table".format(shape_dim_place_table[0],shape_dim_place_table[1]))

We have 18 rows and 4 columns in dim_place_table


In [86]:
dim_time_table = pd.read_sql("SELECT * from covid_19.dim_date", con)
dim_time_table.head()

date_key        date  day  month  year day_of_week  is_holiday
0         1  2020-01-06    6      1  2020      Monday           0
1         2  2020-01-07    7      1  2020     Tuesday           0
2         3  2020-01-08    8      1  2020   Wednesday           0
3         4  2020-01-09    9      1  2020    Thursday           0
4         5  2020-01-10   10      1  2020      Friday           0

In [87]:
# Checking the shape of table
shape_dim_time_table = dim_time_table.shape
print("We have {} rows and {} columns in dim_time_table".format(shape_dim_time_table[0],shape_dim_time_table[1]))

We have 91 rows and 7 columns in dim_time_table


In [88]:
fact_infection_table = pd.read_sql("SELECT * from covid_19.fact_df", con)
fact_infection_table.head()

infection_key  patient_key  place_id  is_released  confirmed_date_key  \
0              1            1         1            1                  18   
1              2            2         1            1                  25   
2              3            3         1            1                  25   
3              4            4         1            1                  25   
4              5            5         1            1                  26   

   released_date_key  visited_1  visited_2  visited_3  infected_by_patient_key  
0                 31          1          1          1                        0  
1                 57          1          5          5                        0  
2                 45          0          0          0                        0  
3                 41          0          0          0                        0  
4                 50          0          0          0                        2

In [89]:
# Checking the shape of table
shape_fact_infection_table = fact_infection_table.shape
print("We have {} rows and {} columns in fact_infection_table".format(shape_fact_infection_table[0],shape_fact_infection_table[1]))

We have 2243 rows and 10 columns in fact_infection_table


## Succes :)

In [90]:
#Query 1
query_1 = '''SELECT D.date,P.province, count(F.infection_key) AS 'Number of infected'
           FROM fact_df as F  join dim_place as P on (F.place_id = P.place_id) join dim_date as D on (F.confirmed_date_key = D.date_key)
           WHERE P.top_3 = 1
           GROUP BY D.date,P.province
           ORDER BY D.date,P.province ; '''
answer1 = pd.read_sql_query(query_1 , con = con)
answer1.head()

date     province  Number of infected
0  2020-01-23        seoul                   1
1  2020-01-26  gyeonggi-do                   1
2  2020-01-27  gyeonggi-do                   1
3  2020-01-30        seoul                   3
4  2020-01-31        seoul                   3

In [91]:
#Query 2
query_2 = '''SELECT D.date,P.province, Sum(F.is_released) AS 'Number of released'
           FROM fact_df as F  join dim_place as P on (F.place_id = P.place_id) join dim_date as D on (F.released_date_key = D.date_key)
           WHERE P.top_3 = 1
           GROUP BY D.date,P.province
           ORDER BY D.date,P.province ; '''

answer2 = pd.read_sql_query(query_2, con = con)
answer2.head()

date     province  Number of released
0  2020-02-05        seoul                 1.0
1  2020-02-09  gyeonggi-do                 1.0
2  2020-02-10        seoul                 1.0
3  2020-02-12  gyeonggi-do                 2.0
4  2020-02-15        seoul                 1.0

In [92]:
#Query 3
query_3 = '''SELECT F.patient_key, sum(F.visited_1) AS 'Place visited in the Last 24 hours',sum(F.visited_2)AS 'Place visited in the Last 48 hours',sum(F.visited_3) AS 'Place visited in the Last 72 hours'
           FROM fact_df AS F JOIN dim_patient AS P on (F.patient_key = P.patient_key) 
           GROUP BY F.patient_key
           Order BY F.patient_key ;'''

answer3 = pd.read_sql_query(query_3, con = con)
answer3.head()

patient_key  Place visited in the Last 24 hours  \
0            1                                 1.0   
1            2                                 1.0   
2            3                                 0.0   
3            4                                 0.0   
4            5                                 0.0   

   Place visited in the Last 48 hours  Place visited in the Last 72 hours  
0                                 1.0                                 1.0  
1                                 5.0                                 5.0  
2                                 0.0                                 0.0  
3                                 0.0                                 0.0  
4                                 0.0                                 0.0

In [93]:
#Query 4
query_4 = '''SELECT count(distinct F.infected_by_patient_key) AS 'Number of Infected Patients over 30'
           FROM fact_df AS F JOIN dim_patient AS P1 on (F.patient_key = P1.patient_key) JOIN dim_patient AS P2 on (F.infected_by_patient_key = P2.patient_key)
           WHERE P2.province = 'seoul' AND P1.age > 30 AND P2.age < 30 ;'''

answer4 = pd.read_sql_query(query_4, con = con)
answer4

Number of Infected Patients over 30
0                                    4

In [94]:
#Query 5
# We calculate the percentage of patients infected with the risk group. When this percentage is higher than 40%.
# We recommend that the provinve will be in quarantine, to protect the population. 
# We assume that the implications of the disease for it can be devastating. 

query_5 = '''SELECT P.province, (count(F.infection_key)) / (P.total_patients) as percent_of_population_in_risk
             FROM fact_df AS F JOIN dim_place AS P on (F.place_id = P.place_id) JOIN dim_patient AS P1 on (F.patient_key = P1.patient_key)
             WHERE P1.age > 55 
             GROUP BY P.province
             HAVING percent_of_population_in_risk > 0.4 ;'''

answer5 = pd.read_sql_query(query_5, con = con)
answer5

province  percent_of_population_in_risk
0              daegu                         0.5873
1  chungcheongbuk-do                         0.4643
2       jeollabuk-do                         0.4286

In [95]:
#Close the connection with MySql
con.close()